src/Fixer/Whitespace/BlankLineBeforeControlStatementFixer.php

```diff
@@ -0,0 +1,265 @@
+                    ]
+                ),
+                new CodeSample(
+                    '<?php
+foreach ($foo as $bar) {
+    if ($bar->isTired()) {
+        $bar->sleep();
+        continue;
+    }
+}',
+                    [
+                        'statements' => ['continue'],
+                    ]
+                ),
+                new CodeSample(
+                    '<?php
+$i = 0;
+do {
+    echo $i;
+} while ($i > 0);
+',
+                    [
+                        'statements' => ['do'],
+                    ]
+                ),
+                new CodeSample(
+                    '<?php
+$a = 9000;
+if (true) {
+    $foo = $bar;
+}',
+                    [
+                        'statements' => ['if'],
+                    ]
+                ),
+                new CodeSample(
+                    '<?php
+
+if (true) {
+    $foo = $bar;
+    return;
+}',
+                    [
+                        'statements' => ['return'],
+                    ]
+                ),
+                new CodeSample(
+                    '<?php
+$a = 9000;
+switch ($a) {
+    case 42:
+        break;
+}',
+                    [
+                        'statements' => ['switch'],
+                    ]
+                ),
+                new CodeSample(
+'<?php
+if (null === $a) {
+    $foo->bar();
+    throw new \UnexpectedValueException("A cannot be null");
+}',
+                    [
+                        'statements' => ['throw'],
+                    ]
+                ),
+                new CodeSample(
+'<?php
+$a = 9000;
+try {
+    $foo->bar();
+} catch (\Exception $exception) {
+    $a = -1;
+}',
+                    [
+                        'statements' => ['try'],
+                    ]
+                ),
+            ],
+            self::$defaultConfiguration
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getPriority()
+    {
+        // should be run after NoUselessReturnFixer
+        return -19;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isAnyTokenKindsFound(array_values($this->fixTokenMap));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        $lineEnding = $this->whitespacesConfig->getLineEnding();
+
+        for ($index = 0, $limit = $tokens->count(); $index < $limit; ++$index) {
+            $token = $tokens[$index];
+
+            if (!$token->isGivenKind(array_values($this->fixTokenMap))) {
+                continue;
+            }
+
+            $prevNonWhitespaceToken = $tokens[$tokens->getPrevNonWhitespace($index)];
+
+            if (!$prevNonWhitespaceToken->equalsAny([';', '}'])) {
+                continue;
+            }
+
+            $prevToken = $tokens[$index - 1];
+
+            if ($prevToken->isWhitespace()) {
+                $parts = explode("\n", $prevToken->getContent());
```

Comment:
note @ self; todo `$parts` is only used for counting, so a line break (substring) count should be enough